### ISSUE TITLE
storage: Coalesced Heartbeats Corner Cases

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#315]
from the discussion in #280, via @bdarnell:

    I think there are some scenarios in which these limited-information heartbeats
    could cause problems. For example, consider a group containing nodes A, B, and C:

    A is initially the leader.
    Node C loses connectivity to the other nodes and after a timeout declares itself a candidate.
    The partition partially heals; C can talk to A but not B.
    C is elected the new leader; A steps down.
    B sees that A has been up continuously and so it still considers A the leader.

    We might be able to solve this on a case-by-case basis (e.g. I think two-round elections would
    help in the above scenario), but to be safe we should probably add a payload to the coalesced
    heartbeat messages to indicate which groups the sending node considers itself to be leader of.

    This thread on raft-dev discusses some related issues:
    https://groups.google.com/forum/?fromgroups#!topic/raft-dev/VgF47vIsezg.
    The term and log index is occasionally useful (but not, I think, strictly required.
    Etcd is not currently sending this information with their heartbeats).

**Comment by tschottdorf** [comment]
The real enemy that I'm worried about is deadlock: A follower with a (stable) stale leader itself doesn't do any harm, except that it won't turn into a candidate. In your example above, with individual heartbeats, B would time out (supposedly after A and C had made progress), become candidate but not receive any votes but its own, repeating indefinitely until the partition heals.

If, somehow, we ended up in a configuration in which there is no leader but every node receives regular heartbeats, that would be a stable configuration we would not appreciate very much.

I suspect that (unless we want to avoid sending the rangeIDs as payload for heartbeats - which at 10^6 consensus groups boils down to a payload of megabytes) we'll be stuck doing tricky things like actively notifying other peers when a leader steps down.

On the other hand, the problem is even more fundamental: At 10^6 ranges, we'll have approximately 10^6/3 leaders, and network issues will cause a huge flood of campaign-related messages. Ultimately the problem is that network issues operate at the level of nodes, but the inflicted traffic scales like the number of ranges. That means we need to keep an eye out on how many ranges a node has - millions of ranges per store means millions of concurrent elections.

Ultimately, the right "unit" for a raft consensus group is not an individual range, but a "node-range" (by which I mean all the ranges which are replicated to the same set of **nodes**). Each such node-range corresponds to a leader, and there would be much fewer reelections etc. in case of connectivity issues (one for each node-range affected by the outage). There are still many possible "node-ranges" (fac(nodes)/(fac(nodes-replicas)*fac(replicas)) but reduced further by datacenter constraints, etc) so on top of that it would make a lot of sense to reuse existing node combinations over wildly going all over the place. Heartbeats could be coalesced by sending the node-ranges as a payload, which should then be feasible size-wise. But all of this paragraph is a deep deep rabbit hole and basically restricts ranges a lot to the point where they're not what they're supposed to be.

I thought about this (again) because when fully implementing read leases (#230), we'll end up with a lot of inter-node communication that needs to happen regularly on a heartbeatInterval time scale, relies on knowing accurately about the leader and, generally, is more complex - much like the campaigning-related messages. So there's another instance coming up where keeping tabs on this is going to be crucial, and I would like to discuss the above issues in this context first.

**Comment by petermattis** [comment]
We're probably going to want to implement something like CopySets (https://www.usenix.org/conference/atc13/technical-sessions/presentation/cidon) which will reduce the number of possible node combinations to use for replication.

**Comment by bdarnell** [comment]
Regarding leaderless deadlock: good point; this is the real problem. There may be a simple solution, though, if we add a response to MsgProp. Right now when a new proposal enters the system, it gets forwarded to the node believed to be the leader, and if that is incorrect, the recipient forwards it again (or drops it if it doesn't know the current leader, but if any node didn't know the current leader it would start an election). If instead of re-forwarding we sent an error back to the originating node, that node could learn that its information is stale and start an election.

Election storms after a network event are definitely going to be a problem. Unfortunately we can't coalesce them the way we can with heartbeats since the vote RPCs have to exchange log positions.

**Comment by tschottdorf** [comment]
@petermattis thanks, that's a good read.
@bdarnell: Good idea. Regarding the network storms, doesn't that mean we should consider something like CopySets (admittedly a much better name than "node-range") as the logical unit on which Raft operates? Splitting and merging (adjacent) ranges shouldn't change drastically, but rebalancing ranges away from a node is somewhat more involved because likely we would have to end up only allowing membership configurations which are a superset of a CopySet (i.e. we can add any node to a CopySet, then remove another node - as long as we end up with a CopySet - so there need to be enough of them or we somehow need to generate new ones - which in turn means having mechanisms to delete some as well). Am I mislead ending up thinking about this venue whenever I consider the multi-raft congestion issues? That will only get worse as we have read leases, which imply (more or less) regular election-like traffic.

**Comment by bdarnell** [comment]
Copysets will be helpful to us, both for minimizing the chance of data loss (which was the original motivation for copysets) and for maximizing the overlap between nodes (so we get the most benefit from coalesced heartbeats). 

Separately, having fewer, larger raft consensus groups will generally make things more efficient, at the expense of coarser load balancing and lower scatter width on recovery (which is also a consequence of copysets). However, the simplest way to have fewer consensus groups is to increase the range split threshold; I'm not sure if having one consensus group manage several ranges is any better than just having bigger ranges.

**Comment by tschottdorf** [comment]
## Improved Leader Awareness for Coalesced Heartbeats

The main issue with coalesced heartbeats is that it is possible that a group gets heartbeat by a group that isn't the current leader any more (and may know about it).

Ideally you would associate with an outgoing heartbeat the Raft groups that it addresses, but that payload would be too large for large installations.

The goal of the procedure below  is to make sure that followers lose their leader once a heartbeat doesn't reach them or the remote leader has stepped down.

Generally it enhances my trust in coalesced heartbeats significantly, and I think in the form below they may be useful to cover all the bases from #326, #357 should we decide that we want to keep coalescing heartbeats for performance and responsability reasons.

- On startup, MultiRaft choses a starting heartbeat index. The main criterion is that over multiple restarts of the instance, the heartbeat index never repeats and increments with every round of coalesced heartbeats sent.

  The wall time of the node at boot time is hence a logical choice, as it's impossible that heartbeats are sent at an average of each nanosecond (in reality, it will be hundreds of milliseconds).
- For each outgoing round of coalesced heartbeats, the `MsgHeartbeat.Index` is set to the current heartbeat index and the index incremented.
- Each heartbeat contains as a payload the local leaders that have stepped down since the last heartbeat (where the members overlap with the respective recipient).
- Upon receipt of such a heartbeat, the host compares the contained heartbeat index with that of the previously received heartbeat from that host (if any). If they are not consecutive numbers, heartbeats must have been lost or reordered.

  The node can thus not fan that heartbeat out to groups, and it can reset the last recorded leader for those groups that consider themselves lead by a group on that host that failed to properly send heartbeats. That means that they won't receive any further heartbeats unless the group has, by means of SoftState on outgoing messages, demonstrated that it has a leader. This relies on the fact that followers never initiate traffic. We should probably ignore the SoftState from outgoing `MsgHeartbeatResp` though.

  The recorded heartbeat index for that host will be reset (so that the next heartbeat is accepted regardless of its index).
- Otherwise, if the heartbeat index is valid, invalidate the leader for those groups from the payload (if any).


The above also opens the door to passing along other information with heartbeats (such as lease information) while guaranteeing that the recipient will immediately fall back to a safe configuration should one of those messages get lost. That would come with a little bit of extra complexity: A missed heartbeat would have to trigger active invalidation messages to the affected Raft groups while above we simply reset the knowledge of the leader in MultiRaft). But it's doable and belongs in another thread.

**Comment by bdarnell** [comment]
What do you mean by "SoftState on outgoing messages"?

**Comment by tschottdorf** [comment]
I mean that it could be a good idea to not update what we know about the leader from `ready.SoftState` unless there is something in `ready.Messages` that isn't a heartbeat response.

Otherwise you could imagine the multiraft resetting leadership for that group, but a delayed heartbeat response to an old coalesced heartbeat setting it back to the stale information right away.

Since the same thing could happen with other message types though, maybe what we really need to do is not interpret SoftState to determine who the range considers their leader when that leader is on a node that at time of inspection is in heartbeat failure state (i.e. for which the last heartbeat led to a reset).

**Comment by bdarnell** [comment]
Ready.SoftState is only present when something has changed; if raft updates the SoftState in response to a heartbeat and we ignore it, we won't get a second update. My question wasn't about ignoring MsgHeartbeatResp, though, it was about the sentence "That means that they won't receive any further heartbeats unless the group has, by means of SoftState on outgoing messages, demonstrated that it has a leader." It sounds like you're basically talking about ignoring SoftState and doing our own inference of leadership based on messages that are sent (I think MsgAppResp is the only one that we can reliably use to determine leadership).

I think this will be prone to deadlock. When raft considers a node to be unresponsive, it stops sending MsgApp to that node, and only sends heartbeats. The MsgApps (or MsgSnaps) will resume only after the leader gets a heartbeat response, so we must allow heartbeats to get through to nodes that are not receiving MsgApp.

**Comment by tschottdorf** [comment]
I did not realize SoftState was only sent on changes, I assumed it was always present. Then that bit of complexity isn't necessary.

With regards to leaders stopping to send `MsgApp` (or snapshots): Good point, I did not have that on the radar. But it doesn't really change anything, does it? `MsgHeartbeatResp`s sent by the groups are ignored by MultiNode anyways, and MultiNode can afford to sent `MsgHeartbeatResp` to all other nodes in general (and typically will anyways). On the leader's side, we won't touch the leadership information, so the heartbeat response will route to the leader as it did before.

Of course we'll have to pull leadership information from outgoing `Msg{Snap,App}` as you suggested, to get the follower back to receiving heartbeats if the group has traffic.

**Comment by bdarnell** [comment]
This thread has mainly been concerned with the problem of a node believing that a given peer is the leader when it is not, and so we solve the problem of deadlocks by clearing a follower's view of the leader. However, we also have problems in the other direction: when a follower believes the group has no leader, it will not fan out coalesced heartbeats and so it will try to trigger an election. With one-phase elections, this is disruptive; with two-phase elections it is wasteful since the node will not be able to update its view of the leader until there is non-heartbeat, non-election activity on the range. We should ensure that the pre-vote RPC includes a way for the node to be informed of the current leader.

**Comment by tschottdorf** [comment]
@bdarnell, this looks ripe to be closed but since there is a reference to the pre-vote RPC please check that there isn't something above that should be carried over to the issue you're likely about to create somewhere.

**Comment by bdarnell** [comment]
I'd like to leave this open until #9380 is done. If coalesced heartbeats include terms for every range (as they do in #9380), then there's nothing to be done here. If we remove per-range information from heartbeats (as in the original implementation of coalesced heartbeats), we'd still need to be careful of this. 

New issue for pre-vote RPC is #9561.

**Comment by arjunravinarayan** [comment]
Closed by #9380, since #9380 includes terms for every range for now.
